impala/hiveserver2.py

```diff
@@ -574,30 +582,38 @@ def next(self):
         return self.__next__()
 
     def __next__(self):
+        self._ensure_buffer_is_filled()
+        log.debug('__next__: popping row out of buffer')
+        return self._buffer.pop()
+
+    def _ensure_buffer_is_filled(self):
         while True:
             if not self.has_result_set:
                 raise ProgrammingError(
                     "Trying to fetch results on an operation with no results.")
             if len(self._buffer) > 0:
-                log.debug('__next__: popping row out of buffer')
-                return self._buffer.pop()
+                return
             elif self._last_operation_active:
-                log.debug('__next__: buffer empty and op is active => fetching '
-                          'more data')
+                log.debug('_fetch_to_buffer_if_needed: buffer empty and op is active '
+                          '=> fetching more data')
                 self._buffer = self._last_operation.fetch(self.description,
                                                           self.buffersize,
                                                           convert_types=self.convert_types)
                 if len(self._buffer) > 0:
-                  log.debug('__next__: popping row out of buffer')
-                  return self._buffer.pop()
+                    return
                 if not self._buffer.expect_more_rows:
-                    log.debug('__next__: no more data to fetch')
+                    log.debug('_ensure_buffer_is_filled: no more data to fetch')
                     raise StopIteration
```

Comment:
Minor: maybe the code would be more straightforward if this method simply returned True/False to signal whether it was successful or not. Raising the exception could be moved to the '\_\_next\_\_' method. 

This way handling 'StopIteration' could be removed from 'fetchall' & 'fetchmany' methods. I think it would make the code clearer as 'fetchall' & 'fetchmany' don't use the cursor's iterator interface, they read from the buffer directly.

Reply:
I am a bit unsure about this. Actually __next__ could also yield instead of returning values and throwing StopIteration at the end. I will do an experiment and benchmark it.

Reply:
I did an experiment and the benchmark was ok, but I would prefer to keep the current solution. One part that became more complex is HiveServer2DictCursor's __next__, because it couldn't just wrap the generator returned by __next__.

Reply:
I would keep it like this for now. There are a lot things that could use a cleanup in impala.cursor.